Helm chart

yuval-k opened the pull request:
Very much draft PR, but submitting to get early feedback:
- Chart structure - does using deployments make sense? the structure of the values.yaml make sense?
- Directory structure: Wasn't sure what the best structure? maybe `install/helm` instead of `kubernetes/helm`?
- I created a `./kubernetes/setup-kind.sh` as a starting point for an end-to-end test. I plan to add an e2e test in the github actions
- Helm unit tests. I'm not familiar with python options; i saw a [declarative one here](https://github.com/quintush/helm-unittest). I never used it before, so wanted to get consensus before i add tests.
- Doc is very much WIP
- Releasing - how do you want this chart released? i see that you currently don't use github releases. This can also be in follow-up PR.

aquamatthias reviewed (commented):
Hey Yuval - thanks for this great start.

To answer your questions:
> Chart structure - does using deployments make sense? the structure of the values.yaml make sense?

Deployments make sense. I added a comment regarding hpa - I think it is not applicable on the current state. Generally I would use args only for command line parameters and not env vars. 

> Directory structure: Wasn't sure what the best structure? maybe install/helm instead of kubernetes/helm?

It looks like a lot of people call it `chart` - so maybe `kubernetes/chart`?

> I created a ./kubernetes/setup-kind.sh as a starting point for an end-to-end test. I plan to add an e2e test in the github actions
Helm unit tests. I'm not familiar with python options; i saw a declarative one here. I never used it before, so wanted to get consensus before i add tests.

Neither did I. But I think it would be great to have a running e2e test. So the effort is very much appreciated.

> Releasing - how do you want this chart released? i see that you currently don't use github releases. This can also be in follow-up PR.

Currently a release of the application is made by [pushing a tag](https://github.com/someengineering/cloudkeeper/blob/main/.github/workflows/docker-build.yml#L4-L8), which then builds a versioned docker image. The chart itself would have its own version, I guess.  We probably need to think about providing a helm repository.


Now to your PR: I did a first pass and tried to follow your instructions and made notes where it did not work for me.
In the end I had a running installation - nice!
There are quite some sharp edges, please find my comments.

So thanks again for this great start 🚀

aquamatthias commented on kubernetes/cloudkeeper/Chart.yaml:
```diff
@@ -0,0 +1,24 @@
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "1.16.0"
```
> i was hoping to avoid having to change the version here as it might introduce maintaince pain.
i figured once we decide on publishing we can publish withhelm package --app-version ... to set it on the fly

Yes that is totally reasonable - it does not have to be the latest version.
Nevertheless: I would set it to something valid - in case the value is not overridden, it would deploy something with a version that can not be fetched and the pods would not start.

yuval-k commented on kubernetes/cloudkeeper/templates/ckcore/deployment.yaml:
```diff
@@ -0,0 +1,80 @@
+            - {{ .Values.ckcore.host | default "0.0.0.0" | quote }}
+            - --start-collect-on-subscriber-connect
+          env:
+          - name: CKCORE_GRAPHDB_SERVER
+            value: {{ .Values.ckcore.graphdb.server  | quote }}
+          - name: CKCORE_GRAPHDB_DATABASE
+            value: {{ .Values.ckcore.graphdb.database | quote }}
+          - name: CKCORE_GRAPHDB_USERNAME
+            value: {{ .Values.ckcore.graphdb.login | quote }}
+          {{- if .Values.ckcore.graphdb.passwordSecret }}
+          - name: CKCORE_GRAPHDB_PASSWORD
```
changed it, except for password, as it comes from a secret

yuval-k commented on docs/source/getting_started/setup_kubernetes.rst:
```diff
@@ -0,0 +1,139 @@
+.. code-block:: bash
+    :caption: Prepare the environment
+
+    kubectl create secret generic cloudkeeper-graphdb-credentials --from-literal=password=$CKCORE_GRAPHDB_PASSWORD
```
good point, i removed the namespace flag form the chart

yuval-k commented on docs/source/getting_started/setup_kubernetes.rst:
```diff
@@ -0,0 +1,139 @@
+
+Configuration
+=============
+Prepare your Helm values file:
```
this file can be created anywhere. i'll clarify that for now (until we release this chart) you need to clone the github repo

aquamatthias reviewed (commented):
Great update! I followed the steps and had a running installation afterwards - great!

I added a couple of minor adjustments, questions and comments, that we should address before merge.
Thanks for the effort!

yuval-k commented on kubernetes/chart/templates/db/arango_deployment.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.db_install }}
+# Install this is a helm hook, so it's not managed by the release.
+# Mainly we want to make sure that this have minimum changes of getting deleted.
+apiVersion: "database.arangodb.com/v1alpha"
+kind: "ArangoDeployment"
+metadata:
+  name: {{ include "cloudkeeper.fullname" . }}-db
+  annotations:
+    helm.sh/hook: pre-install
+    helm.sh/resource-policy: keep
+spec:
+  mode: Single
+{{- end }}
```
we don't; i added this when we thought we can do an easier install by bundling the db CR; i'll remove the `db` folder.

yuval-k commented:
Will address other comments soon

yuval-k commented:
I believe I addressed all comments; let me know if I've missed anything!
is there a way to kick the new ci workflow one time to see that it is indeed working as expected?

lloesche commented:
@yuval-k fantastic work! thank you so much!

The one last thing I would add as a user configurable value is the `PSK` environment variable to all components. It's the same as the `--psk` arg (all components, `ckcore`, `ckmetrics`, `ckworker` and `cksh` take it).

It contains a pre-shared-key (i.e. a passphrase) that is used to derive a 256 bit key that's being used to sign the JWT which authenticates the API requests. That way nobody can just sent API requests to `ckcore`. Long term we want to have proper RBAC and also add TLS to all communication channels. But for the time being we have this pre shared key that all components must know about. The relevant backend code is in https://github.com/someengineering/cloudkeeper/blob/main/cklib/cklib/jwt.py if you're interested in the details.

The chart currently works fine because the container defaults to `PSK=changeme`.

lloesche commented:
> is there a way to kick the new ci workflow one time to see that it is indeed working as expected?

I just ran it. The repo is set to hold workflows from first time contributors. Once we've merged your PR any subsequent PRs of yours would automatically run their workflows.

lloesche commented on kubernetes/chart/templates/ckcore/deployment.yaml:
```diff
@@ -48,6 +48,8 @@ spec:
             - {{ .Values.ckcore.graphdb.database | quote }}
             - --graphdb-username
             - {{ .Values.ckcore.graphdb.login | quote }}
+            - --psk
```
Sorry I meant, the value should be handed in via the environment using the `PSK` env var. The `--psk` arg is just the equivalent to the env var.
Two reasons for this; for one if it's handed in via `--psk` then the secret is visible in the host's process list to anyone with access to it. The other, more important reason, when using `cksh` via the container then there's a special check for `PSK` from the env and a shim is used that provides that PSK to `cksh` transparently. When not handed in via the env  then the user would have to provide `--psk` themselves anytime they call `cksh`.

aquamatthias commented on .github/workflows/k8s-build.yml:
```diff
@@ -0,0 +1,43 @@
+
+jobs:
+    build:
+        name: Build Docker Images
```
```suggestion
        name: Test k8s helm chart
```

aquamatthias reviewed (dismissed):
Great effort. Looks awesome to me - thanks!
Can you take a look into the failing test case? 
Once the test is green, the PR can be merged.

yuval-k commented:
it's hard to know why its failing; i added a describe to the failure block, so hopefully we'll see the reason in this CI run

aquamatthias commented:
@yuval-k mmh still failing. If this helps: we could merge this PR as is and do the fix in a separate PR. Since you are a first time committer, the GitHub Actions do not run automatically atm, but would with the next PR.

yuval-k commented:
yea i knew it would fail, i wanted to see why. good idea - if i can do it in a follow up PR i can iterate this independently

yuval-k commented:
i can disable the check now, so ci will be green

aquamatthias reviewed (approved):
Thanks! Let's fix the GitHub action in a separate PR.

aquamatthias commented:
@yuval-k if you create a PR now, gh actions should run immediately. So you can test on your own pace. 
Hope this helps.
